physx/src/articulation_base.rs

```diff
@@ -75,160 +84,250 @@ impl ArticulationBase {
+            > = Vec::with_capacity(capacity as usize);
+            let len = PxArticulationBase_getLinks(
+                self.as_ptr(),
+                buffer.as_mut_ptr() as *mut *mut _,
+                capacity,
+                0,
+            );
+            buffer.set_len(len as usize);
+            buffer
         }
     }
 
-    #[inline]
-    /// Get an iterator over all links
-    pub fn links(&self) -> &Vec<ArticulationLink> {
-        match self.user_data() {
-            UserData::ArticulationBase(data) => &data.links,
-            _ => unimplemented!(),
+    /// Get a mutable vec of all the links
+    fn get_links_mut(
+        &mut self,
+    ) -> Vec<&mut ArticulationLink<Self::LinkData, Self::ShapeData, Self::MaterialData>> {
+        unsafe {
+            let capacity = PxArticulationBase_getNbLinks(self.as_ptr());
+            let mut buffer: Vec<
+                &mut ArticulationLink<Self::LinkData, Self::ShapeData, Self::MaterialData>,
+            > = Vec::with_capacity(capacity as usize);
+            let len = PxArticulationBase_getLinks(
+                self.as_ptr(),
+                buffer.as_mut_ptr() as *mut *mut _,
+                capacity,
+                0,
+            );
+            buffer.set_len(len as usize);
+            buffer
         }
     }
 
-    #[inline]
-    /// Get an iterator over all links
-    pub fn links_mut(&mut self) -> &mut Vec<ArticulationLink> {
-        match self.user_data_mut() {
-            UserData::ArticulationBase(data) => &mut data.links,
-            _ => unimplemented!(),
-        }
+    /// Get the world bounds of this articulation
+    fn get_world_bounds(&self, inflation: f32) -> PxBounds3 {
+        unsafe { PxArticulationBase_getWorldBounds(self.as_ptr(), inflation).into() }
     }
 
-    /// Create a child link for parent, with the given transform
-    pub fn create_link(
+    /// Add a link to this articulation
+    fn create_link(
         &mut self,
-        parent: PartHandle,
-        transform: Option<Mat4>,
-        joint_transform: Option<Mat4>,
-    ) -> *mut PxArticulationLink {
-        let (parent, trans) = if parent.1 == 0 {
-            (null_mut(), Mat4::identity())
-        } else {
-            let parent_link = parent.1 as *mut PxArticulationLink;
-            (parent_link, unsafe {
-                px_to_gl_tf(PxRigidActor_getGlobalPose(
-                    parent_link as *const PxRigidActor,
-                ))
-            })
-        };
-
-        let transform = joint_transform.unwrap_or_else(Mat4::identity)
-            * transform.unwrap_or_else(Mat4::identity);
-        // eprintln!("creating link with transform: {}", trans * transform);
+        parent: Option<&mut ArticulationLink<Self::LinkData, Self::ShapeData, Self::MaterialData>>,
+        pose: &PxTransform,
+        user_data: Self::LinkData,
+    ) -> Option<Owner<ArticulationLink<Self::LinkData, Self::ShapeData, Self::MaterialData>>> {
```

Comment:
I changed pretty much everything to assoc types.
Good catch on the soundness hole, changed this and the joint creation here to &mut instead, since they aren't added like the actor types, or wholly owned like the foundation/physics/scene objects.